rs/replicated_state/src/replicated_state.rs

```diff
@@ -957,6 +969,80 @@ impl ReplicatedState {
+    ///
+    /// Time complexity: `O(n * log(n))`.
+    pub fn enforce_best_effort_message_limit(&mut self, limit: NumBytes) -> (u64, NumBytes) {
+        const ZERO: NumBytes = NumBytes::new(0);
+
+        // Check if we need to do anything at all before constructing a priority queue.
+        let mut memory_usage = self
+            .canister_states
+            .values()
+            .fold(NumBytes::new(0), |acc, canister| {
+                acc + canister.system_state.best_effort_message_memory_usage()
+            });
+        if memory_usage <= limit {
+            // No need to do anything.
+            return (0, 0.into());
+        }
+
+        // Construct a priority queue of canisters by best-effort message memory usage.
+        let mut priority_queue: BTreeSet<_> = self
+            .canister_states
+            .iter()
+            .filter_map(|(canister_id, canister)| {
+                let memory_usage = canister.system_state.best_effort_message_memory_usage();
+                if memory_usage > ZERO {
+                    Some((memory_usage, *canister_id))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let mut shed_messages = 0;
+        let mut shed_message_bytes = 0.into();
+
+        // Shed messages from the canisters with the largest memory usage until we are
+        // below the limit.
+        while memory_usage > limit && !priority_queue.is_empty() {
+            let Some((memory_usage_before, canister_id)) = priority_queue.pop_last() else {
+                // Safety net, in case a canister somehow reports non-zero best-effort memory
+                // usage but then fails to shed a message.
+                break;
+            };
+
+            // Remove the canister, shed its largest message, replace it.
+            let mut canister = self.canister_states.remove(&canister_id).unwrap();
+            let message_shed = canister
+                .system_state
+                .shed_largest_message(&canister_id, &self.canister_states);
+            debug_assert!(message_shed);
+            let memory_usage_after = canister.system_state.best_effort_message_memory_usage();
+            self.canister_states.insert(canister_id, canister);
+
+            // Replace the canister in the priority queue iff its memory usage is still
+            // non-zero AND a message was actually shed.
+            if memory_usage_after > ZERO && message_shed {
+                priority_queue.insert((memory_usage_after, canister_id));
+            }
+
+            // Update the total memory usage.
+            let memory_usage_delta = memory_usage_before - memory_usage_after;
+            debug_assert!(memory_usage_delta > ZERO);
+            memory_usage -= memory_usage_delta;
+
+            shed_messages += message_shed as u64;
+            shed_message_bytes += memory_usage_delta;
+        }
```

Comment:
Should we have some debug asserts for stats being ok here or do we already have them inside the `shed_largest_message` function?

Reply:
We do.

I did add a (rather expensive, but we don't care) `debug_assert!()` that our running `memory_usage` matches the amount computed by `best_effort_message_memory_taken()`. I guess that's a good safety net against most possibilities of future divergence.